pkg/upload/dsym.go

```diff
@@ -0,0 +1,176 @@
+	Plist       string      `help:"Path to the Info.plist file" type:"path"`
+	ProjectRoot string      `help:"path to remove from the beginning of the filenames in the mapping file" type:"path"`
+	Path        utils.Paths `arg:"" name:"path" help:"Path to directory or file to upload" type:"path" default:"."`
+}
+
+func ProcessDsym(
+	apiKey string,
+	versionName string,
+	scheme string,
+	dev bool,
+	dsymPath string,
+	plistPath string,
+	projectRoot string,
+	paths []string,
+	endpoint string,
+	timeout int,
+	retries int,
+	overwrite bool,
+	dryRun bool,
+) error {
+
+	var buildSettings *ios.XcodeBuildSettings
+	var dsyms *[]*ios.DsymFile
+	var plistData *ios.PlistData
+	var uploadOptions map[string]string
+
+	for _, path := range paths {
+		uploadInfo, err := ios.ProcessPathValue(path, projectRoot)
+		if err != nil {
+			return err
+		}
+
+		// If dsymPath is not set explicitly, use uploadInfo to set it (if available)
+		// uploadInfo.DsymPath is set when <path> is recognised as a dsym path
+		if dsymPath == "" && uploadInfo.DsymPath != "" {
+			dsymPath = uploadInfo.DsymPath
+		}
+
+		// If projectRoot is not set explicitly, use uploadInfo to set it
+		if projectRoot == "" {
+			projectRoot = uploadInfo.ProjectRoot
+		}
+
+		// If scheme is set explicitly, check if it exists
+		if scheme != "" {
+			_, err = ios.IsSchemeInPath(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+
+			// Only when the dsym path is not set, try and work out the scheme
+			if dsymPath == "" {
+				// If the scheme is not set explicitly, try to find it
+				scheme, err = ios.GetDefaultScheme(path, projectRoot)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		log.Info("Using scheme: " + utils.DisplayBlankIfEmpty(scheme))
+
+		// If the dsymPath is not fed in via <path>
+		if dsymPath == "" {
+			buildSettings, err = ios.GetXcodeBuildSettings(path, scheme, projectRoot)
+			if err != nil {
+				return err
+			}
+
+			// Build the dsymPath from build settings
+			dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, buildSettings.DsymName, "Contents", "Resources", "DWARF")
+
+			// Check if dsymPath exists, if not, try alternative path instead
+			err = utils.Path(dsymPath).Validate()
+			if err != nil {
+				log.Info("Could not find dSYM in expected location: " + utils.DisplayBlankIfEmpty(dsymPath))
+				dsymPath = filepath.Join(buildSettings.ConfigurationBuildDir, strings.TrimSuffix(buildSettings.DsymName, ".dSYM"))
+
+				err = utils.Path(dsymPath).Validate()
+				if err != nil {
+					return err
+				}
+				log.Info("Using alternative dSYM path: " + dsymPath)
+			}
+
+		} else {
+			// Use the dsymPath from the command line, check if it's zipped and unzip it if necessary
+			var extractedLocation string
+			extractedLocation, err = utils.ExtractFile(dsymPath, "dsym")
+			if err != nil {
+				log.Warn(utils.DisplayBlankIfEmpty(dsymPath) + " is not a zip file or directory")
```

Comment:
Does `ExtractFile` return `nil` if the target isn't a zip file? If so, when might `err` be set? I was a bit confused by the message (and the fact that it's a Warn) as I wouldn't expect a warning to cause an early return and according to the comments above, we are expecting some cases where it's not a zip file.

Reply:
I've moved all unzip logic to `GetDsymsForUpload` now and I agree this is confusing. If an unsucessful unzip happens we log a warn level log "Could not unzip " + file.Name() + " to a temporary directory, skipping", try to remove any temporary directory that may have gotten created and then keep going with the other uploads.

Reply:
I see – given we'd usually expect just one file I feel like this could be an `Error` message and focussing a bit less on the temp dir (as this isn't something the user knows or can control). The "skipping" bit combined with a warning makes it all seem optional. 

How about:

```
log.Error("Could not unzip " + file.Name() + " for upload")
```

I guess we'd have to update the `log.Error` method to allow a non-exiting error. This makes me think it should honour `--fail-on-upload`, but maybe that's stretching the option too far?

Reply:
I completely agree with the above, and as discussed this flag along with the --ignore-missing-dwarf and --ignore-empty-dsym will be implemented as part of my other PR.